src/Data/Semigroup/Linear.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE StandaloneDeriving #-}
+
+-- | = The linear semigroup hierarchy
+--
+-- TODO: documentation
+
+module Data.Semigroup.Linear
+  ( Semigroup(..)
+  , Monoid(..)
+  , LEndo(..), appLEndo
+  , Top, throw
+  , module Data.Semigroup
+  )
+  where
+
+import Prelude.Linear.Internal.Simple
+import Data.Semigroup hiding (Semigroup(..))
+import qualified Data.Semigroup as Prelude
+import qualified Prelude
+import qualified Unsafe.Linear as Unsafe
+
+class Prelude.Semigroup a => Semigroup a where
+  (<>) :: a ->. a ->. a
+
+class (Semigroup a, Prelude.Monoid a) => Monoid a where
+  {-# MINIMAL #-}
+  mempty :: a
+  mempty = mempty
+  -- convenience redefine, so that nonlinear Data.Semigroup doesn't have
+  -- to be imported to use mempty
+
+---------------
+-- Instances --
+---------------
+
+instance Semigroup () where
+  () <> () = ()
+
+-- | The type of linear endomorphisms.
+newtype LEndo a = LEndo (a ->. a)
+
+-- TODO: have this as a newtype deconstructor once the right type can be
+-- correctly inferred
+appLEndo :: LEndo a ->. a ->. a
+appLEndo (LEndo f) = f
+
+instance Prelude.Semigroup (LEndo a) where
+  LEndo f <> LEndo g = LEndo (f . g)
+instance Prelude.Monoid (LEndo a) where
+  mempty = LEndo id
+instance Semigroup (LEndo a) where
+  LEndo f <> LEndo g = LEndo (f . g)
+instance Monoid (LEndo a) where
+
+instance (Semigroup a, Semigroup b) => Semigroup (a,b) where
+  (a,x) <> (b,y) = (a <> b, x <> y)
+instance (Monoid a, Monoid b) => Monoid (a,b)
+
+instance Semigroup a => Semigroup (Dual a) where
+  Dual x <> Dual y = Dual (y <> x)
+instance Monoid a => Monoid (Dual a)
+
+newtype LWrap a = LWrap a
+  deriving (Prelude.Semigroup, Prelude.Monoid)
+
+-- This instance is unsafe: do not export LWrap so it cannot be used.
+instance Prelude.Semigroup a => Semigroup (LWrap a) where
+  LWrap a <> LWrap b = LWrap (Unsafe.toLinear2 (Prelude.<>) a b)
+instance Prelude.Monoid a => Monoid (LWrap a)
+
+-- XXX: I think these are safe but I'm not fully confident
```

Comment:
No, they unfortunately are not.

Reply:
Does this include the boolean instances below as well? My hope was that even though `LWrap` is unsafe, the simple cases of `Sum` and `Product` would still be safe

Reply:
As long as you don't define them with functions of the `Unsafe` module, the boolean instances are safe.